lib/snapshot/snapshot_repository.rb

```diff
@@ -0,0 +1,134 @@
+require 'elasticsearch'
+require "logging"
+require "json"
+
+module Snapshot
+  class SnapshotRepositoryBusy < Exception
+  end
+
+  class SnapshotRepository
+    attr_reader :repository_name
+
+    def initialize(base_uri:, repository_name:)
+      @client = Elasticsearch::Client.new(host: base_uri, log: true).snapshot
+      @repository_name = repository_name
+    end
+
+    # Create a snapshot repository
+    # The verify param requires delete permissions on the underlying data store
+    def create_repository(type, settings)
+      @client.create_repository(
+        repository: "#{repository_name}",
+        body: {
+          type: type,
+          settings: settings
+        }
+      )
+    end
+
+    # Start a snapshot of one or more indices.
+    # snapshot_name is assumed to be unique.
+    #
+    # Raises a ServiceUnavailable exception if elasticsearch is already taking
+    # another snapshot.
+    def snapshot(snapshot_name, index_names)
+      @client.create(
+        repository: repository_name,
+        snapshot: snapshot_name,
+        body: {
+          indices: index_names.join(","),
+          include_global_state: false
+        }
+      )
+    end
+
+    # Get the status of a previously started snapshot
+    def check_snapshot(snapshot_name)
+      response = @client.status(
+        repository: repository_name,
+        snapshot: snapshot_name,
+      )
+      snapshot = response["snapshots"].first
+
+      logger.info(
+        "Snapshot #{snapshot['snapshot']}: #{snapshot['state']}"
+      )
+      logger.debug(snapshot["shards_stats"].to_s)
+
+      snapshot["state"]
+    end
+
+    def in_progress_snapshots
+      snapshots = @client.status(repository: repository_name)["snapshots"]
+      snapshots.reject! { |snapshot| snapshot.fetch("state") == "SUCCESS" }
+      snapshots.map { |snapshot| snapshot.fetch("snapshot") }
+    end
+
+    def last_successful_snapshot(snapshots)
+      (snapshots - in_progress_snapshots).last
+    end
+
+    def restore_indexes(snapshot_name, index_groups)
```

Comment:
I think so. `#select_index_from_groups` `guess_restored_indices` and `restore_indexes` are all part of the restore process so I think we can move these to a helper object.

I was toying with the idea of having a Snapshot as a separate thing as well but don't think that would help here.